pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java

```diff
@@ -584,6 +584,18 @@ public boolean setTransactionTimeout(int seconds) {
     return false;
   }
 
+  private int mapSQLStateToXAErrorCode(SQLException sqlException) {
+    if (isPostgreSQLIntegrityConstraintViolation(sqlException)) {
```

Comment:
Is there any reason why integrity constraint violation exceptions only are treated special?

Reply:
These exceptions can be mapped to the correct XA Error Code. When Postgres reported an integrity constraint violation that means the change has been rolled back and can be reported as such. We can not make the same claim for other exceptions.